01_fib/jimbotech/ex_fibonacci.go

```diff
@@ -0,0 +1,47 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"math"
+	"os"
+)
+
+var out io.Writer = os.Stdout
+
+// fibonacci returns a function that returns
+// an number in the fibonacci sequence.
+// This is an implmentation of "closure".
+// The inside function returns the current number
+// in the fibonnacci sequence, starting at 1
+// and sets up the next which will be returned
+// in the subsequent call
+func fibonacci() func() int {
+	secondLast, last := 0, 1
+
+	return func() int {
+		result := last
+		secondLast, last = last, secondLast+last
+		return result
+	}
+}
+
+func fib(n int) {
+	f := fibonacci()
+
+	if n < 0 {
+		for i := 0; i > n; i-- {
+			factor := int(math.Pow(-1, float64(i*-1)))
```

Comment:
The computation of negative fibonacci numbers really belongs where the sequence is calculated. This function is responsible for printing `n` fib numbers. Instead you've split the computation over two separate functions for negative numbers. For instance, if you wanted to have another function return a slice of fib numbers instead of printing them, you would have to duplicate this negative number handling. But you've kind of backed yourself into a corner with your closure implementation of fib - it has no notion of which direction to go in so it cannot compute negative numbers.

Also the computation of -1 or 1 here is a bit obtuse. `i*-1` is more simply expressed as `-i` and using floating point calculations just to alternate between -1 and 1 seems unnecessary. If it were possible to write it as `-1^-i` in simple mathematical notation, then it may be worth doing. But with the casts and `math.Pow` instead of `^` or similar, it becomes that much harder to mentally parse and understand. An alternative is to use a variable `sign` initialised to `1`, and each time around the loop you just do `sign *= -1` - that way it alternates but is a much simpler expression.

Reply:
Done ... however I feel hardly done by. I feel I am handling the negative numbers in the right place as the series is being generated there. So now I have separated the generation from the printing. Am I going to be criticised for inefficiency now? I was hoping for bonus points not bonus corrections. Secondly, as this is a exerciser I felt proud of my play with the math library and expected applause. Yes I did not like floating point either. When I re-wrote it I needed ABS, guess what, floating point again. So can I use any of the math library?